types/src/transaction/transaction.rs

```diff
@@ -346,6 +317,109 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
     sum.iter().all(|(_, sum)| sum.is_zero())
 }
 
+fn apply_bitmask_to_output(
+    mut bitmask: Vec<u8>,
+    filter_len: usize,
+    outputs: Vec<AssetTransferOutput>,
+    mut result: Vec<AssetTransferOutput>,
+) -> Result<Vec<AssetTransferOutput>, HashingError> {
+    let mut index = 0;
+    let output_len = outputs.len();
+
+    loop {
+        if !(index < filter_len) {
+            return Ok(result)
+        }
+
+        let mut filter = bitmask.pop().unwrap();
+
+        for i in 0..8 {
+            if index == output_len as usize {
+                return Ok(result)
+            }
+
+            if (filter & 0x1) == 1 {
+                result.push(outputs[index + i].clone());
+            }
+
+            filter = filter >> 1;
+        }
+        index += 1;
+    }
+}
+
+fn is_sign_all(
+    inputs: &Vec<AssetTransferInput>,
+    scheme: bool,
+    sign_single: bool,
+    cur: &AssetOutPoint,
+) -> Vec<AssetTransferInput> {
+    if scheme {
+        return inputs
+            .iter()
+            .map(|input| AssetTransferInput {
+                prev_out: input.prev_out.clone(),
+                lock_script: Vec::new(),
+                unlock_script: Vec::new(),
+            })
+            .collect()
+    }
+
+    if sign_single {
+        return vec![AssetTransferInput {
+            prev_out: cur.clone(),
+            lock_script: Vec::new(),
+            unlock_script: Vec::new(),
+        }]
+    }
+
+    Vec::new()
+}
+
+impl PartialHashing for Transaction {
+    fn hash_partially(&self, mut bitvec: Vec<u8>, cur: &AssetOutPoint, is_burn: bool) -> Result<H256, HashingError> {
+        let tag = bitvec.pop().unwrap();
+        let sign_all_inputs = (tag & 0x1) == 1;
+        let sign_all_outputs = (tag >> 1 & 0x1) == 1;
+        let filter_len = (tag >> 2) & 0x3f;
+
+        match self {
+            Transaction::AssetTransfer {
+                network_id,
+                burns,
+                inputs,
+                outputs,
+                nonce,
+            } => {
+                let new_burns = is_sign_all(burns, sign_all_inputs, is_burn, cur);
+                let new_inputs = is_sign_all(inputs, sign_all_inputs, !is_burn, cur);
+
+                let new_outputs = if sign_all_outputs {
+                    outputs.clone()
+                } else {
+                    if bitvec.len() != filter_len as usize {
+                        return Err(HashingError::InvalidFilter)
+                    }
+                    apply_bitmask_to_output(bitvec.clone(), filter_len as usize, outputs.clone(), Vec::new())?
```

Comment:
What happens if the length of `outputs` is not the same with `bitvec.len()`

Reply:
I defined some rules.

1. filter_len must be same with the bitvec's byte length which is the tag except for one byte for output scheme, input scheme and filter_len.
2. filter_len can be greater than output's length. Only exist outputs will be signed. So an output can be added if a signer wants to do. He just adds zero behind the filter. If he doesn't want to extend the outputs, he can add 1 behind the filter.
3. filter_len can be shorter than output's length. Only relevant outputs will be signed.

I am now writing the specification for partial signing. If you give me your opinion, I will summarize that with the above rules.